subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/r68/KotlinDslScriptsModelCrossVersionSpec.groovy

```diff
@@ -0,0 +1,232 @@
+
+        when:
+        def model = kotlinDslScriptsModelFor()
+
+        then:
+        model.scriptModels.keySet() == spec.scripts.values() as Set
+
+        and:
+        assertModelMatchesBuildSpec(model, spec)
+    }
+
+    def "can fetch model for the scripts of a build in lenient mode"() {
+
+        given:
+        def spec = withBuildSrcAndInitScripts()
+
+        and:
+        spec.scripts.init << """
+            script_body_compilation_error
+        """
+
+        when:
+        def model = kotlinDslScriptsModelFor(true)
+
+        then:
+        model.scriptModels.keySet() == spec.scripts.values() as Set
+
+        and:
+        assertModelMatchesBuildSpec(model, spec)
+
+        and:
+        assertHasExceptionMessage(
+            model,
+            spec.scripts.init,
+            "Unresolved reference: script_body_compilation_error"
+        )
+    }
+
+    def "can fetch model for a given set of scripts"() {
+
+        given:
+        def spec = withBuildSrcAndInitScripts()
+        def requestedScripts = spec.scripts.values()
+
+        when:
+        def model = kotlinDslScriptsModelFor(requestedScripts)
+
+        then:
+        model.scriptModels.keySet() == requestedScripts as Set
+
+        and:
+        assertModelMatchesBuildSpec(model, spec)
+    }
+
+    def "can fetch model for a given set of scripts of a build in lenient mode"() {
+
+        given:
+        def spec = withBuildSrcAndInitScripts()
+        def requestedScripts = spec.scripts.values()
+
+        and:
+        spec.scripts.init << """
+            script_body_compilation_error
+        """
+
+        when:
+        def model = kotlinDslScriptsModelFor(true, requestedScripts)
+
+        then:
+        model.scriptModels.keySet() == requestedScripts as Set
+
+        and:
+        assertModelMatchesBuildSpec(model, spec)
+
+        and:
+        assertHasExceptionMessage(
+            model,
+            spec.scripts.init,
+            "Unresolved reference: script_body_compilation_error"
+        )
+    }
+
+    def "single request models equal multi requests models"() {
+
+        given:
+        def spec = withBuildSrcAndInitScripts()
+
+        when:
+        Map<File, KotlinDslScriptModel> singleRequestModels = kotlinDslScriptsModelFor().scriptModels
+
+        and:
+        Map<File, KotlinBuildScriptModel> multiRequestsModels = spec.scripts.values().collectEntries {
+            [(it): kotlinBuildScriptModelFor(projectDir, it)]
+        }
+
+        then:
+        spec.scripts.values().each { script ->
+            assert singleRequestModels[script].classPath == multiRequestsModels[script].classPath
+            assert singleRequestModels[script].sourcePath == multiRequestsModels[script].sourcePath
+            assert singleRequestModels[script].implicitImports == multiRequestsModels[script].implicitImports
+        }
+    }
+
+    private KotlinDslScriptsModel kotlinDslScriptsModelFor(boolean lenient = false, File... scripts) {
```

Comment:
:x: This function and most of the ones below are now duplicated with `r60/KotlinDslScriptsModelCrossVersionSpec.groovy`. They should be extracted to a base class.